libsql/src/hrana/stream.rs

```diff
@@ -0,0 +1,375 @@
+where
+    T: for<'a> HttpSend<'a>,
+{
+    fn clone(&self) -> Self {
+        HttpStream {
+            inner: self.inner.clone(),
+        }
+    }
+}
+
+impl<T> HttpStream<T>
+where
+    T: for<'a> HttpSend<'a>,
+{
+    pub(super) fn open(client: T, base_url: String, auth_token: String) -> Self {
+        HttpStream {
+            inner: Arc::new(Inner {
+                affected_row_count: AtomicU64::new(0),
+                last_insert_rowid: AtomicI64::new(0),
+                stream: Mutex::new(RawStream {
+                    client,
+                    base_url,
+                    auth_token,
+                    sql_id_generator: 0,
+                    status: StreamStatus::Open,
+                    baton: None,
+                }),
+            }),
+        }
+    }
+
+    /// Executes a final request and immediately closes current stream - all in one request.
+    pub async fn finalize(&mut self, req: StreamRequest) -> Result<StreamResponse> {
+        let mut client = self.inner.stream.lock().await;
+        let resp = client.finalize(req).await?;
+        Ok(resp)
+    }
+
+    pub async fn close(&mut self) -> Result<()> {
+        let mut client = self.inner.stream.lock().await;
+        client.close().await
+    }
+
+    pub async fn execute(&self, stmt: Stmt) -> Result<StmtResult> {
+        let mut client = self.inner.stream.lock().await;
+        let resp = client
+            .send(StreamRequest::Execute(ExecuteStreamReq { stmt }))
+            .await?;
+        match resp {
+            StreamResponse::Execute(resp) => {
+                let r = resp.result;
+                self.inner
+                    .affected_row_count
+                    .store(r.affected_row_count, Ordering::SeqCst);
+                self.inner
+                    .last_insert_rowid
+                    .store(r.last_insert_rowid.unwrap_or_default(), Ordering::SeqCst);
+                Ok(r)
+            }
+            other => unexpected!(other),
+        }
+    }
+
+    pub async fn batch(&self, batch: Batch) -> Result<BatchResult> {
+        let mut client = self.inner.stream.lock().await;
+        let resp = client
+            .send(StreamRequest::Batch(BatchStreamReq { batch }))
+            .await?;
+        match resp {
+            StreamResponse::Batch(resp) => Ok(resp.result),
+            other => unexpected!(other),
+        }
+    }
+
+    pub async fn store_sql(&self, sql: String) -> Result<StoredSql<T>> {
+        let mut client = self.inner.stream.lock().await;
+        let sql_id = client.next_sql_id();
+        let resp = client
+            .send(StreamRequest::StoreSql(StoreSqlStreamReq { sql, sql_id }))
+            .await?;
+        match resp {
+            StreamResponse::StoreSql => {
+                drop(client);
+                Ok(StoredSql::new(self.clone(), sql_id))
+            }
+            other => {
+                client.sql_id_generator = client.sql_id_generator.wrapping_sub(1);
```

Comment:
I'm not entirely sure about that, can we trust that the server hasn't stored the statement, despite returning an error? Wouldn't it be safe to just discard this sql_id altogether?